### PR TITLE
TST: instrument test class update

### PR DESCRIPTION
**Pull request by jklenzing**
# Description

Addresses #518 

- For instrument end-to-end testing, stores each instrument as a `dict` rather than a `pysat.Instrument` object to aid in debugging
- Removes `try`/`except` wrappers from tests that were used to improve this debugging blockade previously
- Renames `package` variable as `inst_loc` to improve readability
- Consistent use of `inst_name` for the full instrument name, ie `pysat_testing`
- Uses cleanup methods in `tempfile` rather than removing files separately
- Adds tests to ensure `generate_instrument_list` behaves as expected if broken instruments are in a subpackage

## Type of change

- Testing fix (non-breaking change which fixes an issue)
- Breaking change (fix or feature that would cause existing functionality
      to not work as expected)
- This change requires a documentation update

# How Has This Been Tested?

Tested locally for both the core package and the `maint/inst_test_updates` branch of the pysatNASA package.  
Run the following:
```
pytest -vs test_instruments.py
```
in both the `pysat/tests`directory and the `pysatNASA/tests` directory.

**Test Configuration**:
* Mac OS X 10.15.6
* Python 3.8.2

# Checklist:

- [x] Make sure you are merging into the ``develop`` (not ``main``) branch
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [na] Add a note to ``CHANGELOG.md``, summarizing the changes -- changes part of instrument test_class, first implemented in this version

### CODE REVIEW

**Comment by jklenzing** [comment]
NOTE: this will require changes in all downstream packages that use this test_class.  Starting with pysatNASA as this engages all the tests available, will move to the others once we finalize syntax here.
- [ ] pysatCDAAC
- [x] pysatNASA
- [ ] pysatIncubator
- [ ] pysatMadrigal
- [ ] pysatMissions
- [ ] pysatModels
- [ ] pysatSpaceWeather

**Comment by jklenzing** [comment]
Note: while comparing the output before and after this pull, it seems that the changes to `repr` in #547 have solved the transparency issues discussed previously.  The switch to dictionaries rather than inst objects implemented here is not strictly needed now, but is likely cleaner.  Open to discussion.

**Review by aburrell — COMMENTED** [review]
Code looks good to me.  Where do you recommend I try this out?

**Comment by jklenzing** [comment]
You can download the test branch of pysatNASA and run the instrument tests locally.  Likewise, making updates to the syntax to any of the packages based on the changes in pysat/pysatNASA#10 should work as a test case.

**Comment by jklenzing** [comment]
Looking over coveralls, there's a few parts where some of the robust checks are not covered.  I'm going to add in a few checks to confirm this is working as it should.

**Review by rstoneback — COMMENTED** [review]
Thanks @jklenzing!

**Comment by rstoneback** [comment]
> Note: while comparing the output before and after this pull, it seems that the changes to `repr` in #547 have solved the transparency issues discussed previously. 

Nice!

File: docs/dependency.rst [hunk 1]
```diff
@@ -116,17 +121,18 @@ generally be unchanged.
...
               getattr(InstTestClass, method).pytestmark.append(mark)
           elif 'no_download' in names:
-              mark = pytest.mark.parametrize("inst", instruments['no_download'])
+              mark = pytest.mark.parametrize("inst_dict",
+                                             instruments['no_download'])
               getattr(InstTestClass, method).pytestmark.append(mark)
 
 Finally, the ``setup`` function under the ``TestInstruments`` class should be
-updated with the package name.
+updated with the location of the instrument library.
```
[the remark this answers is not in the document]
Comment (jklenzing) [reply docs/dependency.rst:135]: I'm fine with calling it a sub package

**Comment by aburrell** [comment]
So, I tested this by creating a new pysatMadrigal branch were I updated things: https://github.com/pysat/pysatMadrigal/tree/inst_test_develop-3

Previously, all of the unit tests passed.  With this PR and the new Madrigal branch, the following failures occur.

```
=================================================== test session starts ===================================================
platform darwin -- Python 3.7.9, pytest-6.0.1, py-1.9.0, pluggy-0.13.1
rootdir: /Users/aburrell/Programs/Git/pysatMadrigal, configfile: setup.cfg
plugins: flake8-1.0.4, cov-2.10.0
collected 105 items                                                                                                       

test_instruments.py FFFFFF......F.........F.................................sssssssssss                             [ 63%]
test_utils_coords.py ......................................                                                         [100%]

======================================================== FAILURES =========================================================
_____________________________________ TestInstruments.test_modules_standard[dmsp_ivm] _____________________________________

self = <pysatMadrigal.tests.test_instruments.TestInstruments object at 0x1115d6dd0>, inst_name = 'dmsp_ivm'

    @pytest.mark.all_inst
    def test_modules_standard(self, inst_name):
        """Checks that modules are importable and have standard properties.
        """
        # ensure that each module is at minimum importable
        module = import_module(''.join(('.', inst_name)),
>                              package=self.inst_loc.__name__)
E       AttributeError: 'TestInstruments' object has no attribute 'inst_loc'

../../../pysat/pysat/tests/instrument_test_class.py:214: AttributeError
_____________________________________ TestInstruments.test_modules_standard[jro_isr] ______________________________________

self = <pysatMadrigal.tests.test_instruments.TestInstruments object at 0x111f203d0>, inst_name = 'jro_isr'

    @pytest.mark.all_inst
    def test_modules_standard(self, inst_name):
        """Checks that modules are importable and have standard properties.
        """
        # ensure that each module is at minimum importable
        module = import_module(''.join(('.', inst_name)),
>                              package=self.inst_loc.__name__)
E       AttributeError: 'TestInstruments' object has no attribute 'inst_loc'

../../../pysat/pysat/tests/instrument_test_class.py:214: AttributeError
________________________________ TestInstruments.test_standard_function_presence[dmsp_ivm] ________________________________

self = <pysatMadrigal.tests.test_instruments.TestInstruments object at 0x1117527d0>, inst_name = 'dmsp_ivm'

    @pytest.mark.all_inst
    def test_standard_function_presence(self, inst_name):
        """Check if each function is callable and all required functions exist
        """
        module = import_module(''.join(('.', inst_name)),
>                              package=self.inst_loc.__name__)
E       AttributeError: 'TestInstruments' object has no attribute 'inst_loc'

../../../pysat/pysat/tests/instrument_test_class.py:246: AttributeError
________________________________ TestInstruments.test_standard_function_presence[jro_isr] _________________________________

self = <pysatMadrigal.tests.test_instruments.TestInstruments object at 0x111f20590>, inst_name = 'jro_isr'

    @pytest.mark.all_inst
    def test_standard_function_presence(self, inst_name):
        """Check if each function is callable and all required functions exist
        """
        module = import_module(''.join(('.', inst_name)),
>                              package=self.inst_loc.__name__)
E       AttributeError: 'TestInstruments' object has no attribute 'inst_loc'

../../../pysat/pysat/tests/instrument_test_class.py:246: AttributeError
__________________________________ TestInstruments.test_instrument_test_dates[dmsp_ivm] ___________________________________

self = <pysatMadrigal.tests.test_instruments.TestInstruments object at 0x111e22f10>, inst_name = 'dmsp_ivm'

    @pytest.mark.all_inst
    def test_instrument_test_dates(self, inst_name):
        """Check that module has structured test dates correctly."""
        module = import_module(''.join(('.', inst_name)),
>                              package=self.inst_loc.__name__)
E       AttributeError: 'TestInstruments' object has no attribute 'inst_loc'

../../../pysat/pysat/tests/instrument_test_class.py:261: AttributeError
___________________________________ TestInstruments.test_instrument_test_dates[jro_isr] ___________________________________

self = <pysatMadrigal.tests.test_instruments.TestInstruments object at 0x1120080d0>, inst_name = 'jro_isr'

    @pytest.mark.all_inst
    def test_instrument_test_dates(self, inst_name):
        """Check that module has structured test dates correctly."""
        module = import_module(''.join(('.', inst_name)),
>                              package=self.inst_loc.__name__)
E       AttributeError: 'TestInstruments' object has no attribute 'inst_loc'

../../../pysat/pysat/tests/instrument_test_class.py:261: AttributeError
________________________________________ TestInstruments.test_download[inst_dict6] ________________________________________

self = <pysatMadrigal.tests.test_instruments.TestInstruments object at 0x111f82290>
inst_dict = {'inst_module': <module 'pysatMadrigal.instruments.jro_isr' from '/Users/aburrell/Programs/Git/pysatMadrigal/pysatMadrigal/instruments/jro_isr.py'>, 'sat_id': '', 'tag': 'drifts_ave'}

    @pytest.mark.first
    @pytest.mark.download
    def test_download(self, inst_dict):
        """Check that instruments are downloadable."""
    
        test_inst, date = initialize_test_inst_and_date(inst_dict)
    
        # check for username
        inst_name = '_'.join((test_inst.platform, test_inst.name))
        dl_dict = user_download_dict[inst_name] if inst_name in \
            user_download_dict.keys() else {}
>       test_inst.download(date, date, **dl_dict)

../../../pysat/pysat/tests/instrument_test_class.py:278: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../../../pysat/pysat/_instrument.py:1786: in download
    **kwargs)
../instruments/jro_isr.py:164: in download
    data_path=data_path, user=user, password=password)
../instruments/methods/madrigal.py:244: in download
    "pysat", format=file_format)
/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/madrigalWeb/madrigalWeb.py:1770: in downloadFile
    urlFile = urllib2.urlopen(url)
/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/urllib/request.py:222: in urlopen
    return opener.open(url, data, timeout)
/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/urllib/request.py:531: in open
    response = meth(req, response)
/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/urllib/request.py:641: in http_response
    'http', request, response, code, msg, hdrs)
/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/urllib/request.py:563: in error
    result = self._call_chain(*args)
/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/urllib/request.py:503: in _call_chain
    result = func(*args)
/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/urllib/request.py:755: in http_error_302
    return self.parent.open(new, timeout=req.timeout)
/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/urllib/request.py:531: in open
    response = meth(req, response)
/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/urllib/request.py:641: in http_response
    'http', request, response, code, msg, hdrs)
/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/urllib/request.py:569: in error
    return self._call_chain(*args)
/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/urllib/request.py:503: in _call_chain
    result = func(*args)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <urllib.request.HTTPDefaultErrorHandler object at 0x111f20190>, req = <urllib.request.Request object at 0x111f8f650>
fp = <http.client.HTTPResponse object at 0x111f8fc90>, code = 504, msg = 'Gateway Timeout'
hdrs = <http.client.HTTPMessage object at 0x111f8ff90>

    def http_error_default(self, req, fp, code, msg, hdrs):
>       raise HTTPError(req.full_url, code, msg, hdrs, fp)
E       urllib.error.HTTPError: HTTP Error 504: Gateway Timeout

/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/urllib/request.py:649: HTTPError
-------------------------------------------------- Captured stderr call ---------------------------------------------------
pysat WARNING: Unable to find any files that match the supplied template. If you have the necessary files please check pysat settings and file locations (e.g. pysat.pysat_dir).
---------------------------------------------------- Captured log call ----------------------------------------------------
WARNING  pysat:_files.py:387 Unable to find any files that match the supplied template. If you have the necessary files please check pysat settings and file locations (e.g. pysat.pysat_dir).
_______________________________________ TestInstruments.test_load[none-inst_dict6] ________________________________________

self = <pysatMadrigal.tests.test_instruments.TestInstruments object at 0x113554f50>, clean_level = 'none'
inst_dict = {'inst_module': <module 'pysatMadrigal.instruments.jro_isr' from '/Users/aburrell/Programs/Git/pysatMadrigal/pysatMadrigal/instruments/jro_isr.py'>, 'sat_id': '', 'tag': 'drifts_ave'}

    @pytest.mark.second
    @pytest.mark.download
    @pytest.mark.parametrize("clean_level", ['none', 'dirty', 'dusty', 'clean'])
    def test_load(self, clean_level, inst_dict):
        """Check that instruments load at each cleaning level."""
    
        test_inst, date = initialize_test_inst_and_date(inst_dict)
        if len(test_inst.files.files) > 0:
            # Set Clean Level
            test_inst.clean_level = clean_level
            target = 'Fake Data to be cleared'
            test_inst.data = [target]
            try:
                test_inst.load(date=date)
            except ValueError as verr:
                # Check if instrument is failing due to strict time flag
                if str(verr).find('Loaded data') > 0:
                    test_inst.strict_time_flag = False
                    with warnings.catch_warnings(record=True) as war:
                        test_inst.load(date=date)
                    assert len(war) >= 1
                    categories = [war[j].category for j in range(0, len(war))]
                    assert UserWarning in categories
                else:
                    # If error message does not match, raise error anyway
                    raise(verr)
    
            # Make sure fake data is cleared
            assert target not in test_inst.data
            # If cleaning not used, something should be in the file
            # Not used for clean levels since cleaning may remove all data
            if clean_level == "none":
>               assert not test_inst.empty
E               AssertionError

../../../pysat/pysat/tests/instrument_test_class.py:313: AssertionError
==================================================== warnings summary =====================================================
pysatMadrigal/tests/test_instruments.py::TestInstruments::test_load[clean-inst_dict5]
  /Users/aburrell/Programs/Git/pysat/pysat/tests/instrument_test_class.py:45: UserWarning: Files > 20.  Not deleted.  Please check to ensure temp directory is used
    warnings.warn(''.join(('Files > 20.  Not deleted.  Please check to ',

-- Docs: https://docs.pytest.org/en/stable/warnings.html
================================================= short test summary info =================================================
FAILED test_instruments.py::TestInstruments::test_modules_standard[dmsp_ivm] - AttributeError: 'TestInstruments' object ...
FAILED test_instruments.py::TestInstruments::test_modules_standard[jro_isr] - AttributeError: 'TestInstruments' object h...
FAILED test_instruments.py::TestInstruments::test_standard_function_presence[dmsp_ivm] - AttributeError: 'TestInstrument...
FAILED test_instruments.py::TestInstruments::test_standard_function_presence[jro_isr] - AttributeError: 'TestInstruments...
FAILED test_instruments.py::TestInstruments::test_instrument_test_dates[dmsp_ivm] - AttributeError: 'TestInstruments' ob...
FAILED test_instruments.py::TestInstruments::test_instrument_test_dates[jro_isr] - AttributeError: 'TestInstruments' obj...
FAILED test_instruments.py::TestInstruments::test_download[inst_dict6] - urllib.error.HTTPError: HTTP Error 504: Gateway...
FAILED test_instruments.py::TestInstruments::test_load[none-inst_dict6] - AssertionError
============================= 8 failed, 86 passed, 11 skipped, 1 warning in 503.06s (0:08:23) =============================
```

**Comment by jklenzing** [comment]
Self.package needs to be updated to self.inst_loc in the setup and teardown

**Comment by jklenzing** [comment]
https://github.com/pysat/pysatMadrigal/blob/5f53f7a155f00a44f3b08ad0a44af2182454a56a/pysatMadrigal/tests/test_instruments.py#L37-L47

**Comment by aburrell** [comment]
That's the type of bug I like... something that is easily corrected by a second set of eyes :P.

**Comment by jklenzing** [comment]
OK, I've finished adding the tests on the instrument generation routine.  They are currently under test_instrument.py (won't be collected if in the instrument test class file).

The other thing we've discussed is moving the `generate_instrument_list` function to somewhere else in the code.  I would lean toward placing this under the core package, especially if we build off of this for a user interface.  Thoughts?

**Comment by aburrell** [comment]
Ok, Madrigal tests now passing locally on the specified branch.  It looks like issues were introduced in travis, though :(

**Comment by jklenzing** [comment]
Madrigal should fail in travis unless you've updated the pysat branch there

**Comment by aburrell** [comment]
> The other thing we've discussed is moving the generate_instrument_list function to somewhere else in the code. I would lean toward placing this under the core package, especially if we build off of this for a user interface. Thoughts?

I think core is the right place.

**Comment by aburrell** [comment]
> Madrigal should fail in travis unless you've updated the pysat branch there

I was referring to the pysat travis tests.  Which you fixed.

**Comment by jklenzing** [comment]
As a headsup, I found a bug where the unit tests in this test version were not properly using the temp directories.  You may need to re-download files for the test days.

**Comment by jklenzing** [comment]
This one should be ready to go.  I've fixed the use of `tempfile`, as well as removed the `remove_files` method from the tests so that system mods are done in a safer fashion.  Note that the tests now have setup and teardown at both the class and method level since the load tests require instruments to be in the same location as the download tests.  The pysatNASA branch is working locally, code has been updated there as a template (in addition to the test_instruments.py file here).

File: pysat/utils/_core.py [hunk 2]
```diff
@@ -1,9 +1,8 @@
-from __future__ import print_function
-from __future__ import absolute_import
-
+import datetime as dt
+from importlib import import_module
+from importlib import reload as re_load
```
Comment (aburrell) [inline pysat/utils/_core.py:3]: Why the nickname?

**Review by aburrell — COMMENTED** [review]
Going to see if this still works on the Madrigal branch.

File: pysat/utils/_core.py (hunk 2 above)
Comment (jklenzing) [reply pysat/utils/_core.py:3]: inherited from the python 2/3 compatibility days.  I could fix it here or we could do a larger scrub later.

Comment (aburrell) [reply pysat/utils/_core.py:3]: I thought that might be it.  I'd recommend just having `import importlib` now that we aren't supporting Python 2.

Comment (aburrell) [reply pysat/utils/_core.py:3]: And I am a fan of fixing the files as we go, if you're cool with that.

Comment (jklenzing) [reply pysat/utils/_core.py]: fixed here and in test_instrument.py

**Review by rstoneback — APPROVED** [review]
Thanks @jklenzing. Changes look good to me. Approving my part - not jumping ahead of @aburrell.

File: pysat/tests/test_instrument.py [hunk 3]
```diff
@@ -1627,3 +1628,43 @@ def setup(self):
...
+            assert 'broken_inst' not in dict['inst_module'].__name__
+
+    def test_for_missing_test_date(self):
+        """Check that instruments without _test_dates are still added to the list
+        """
+        del self.test_library.pysat_testing._test_dates
+        # If an instrument does not have the _test_dates attribute, it should
+        # still be added to the list for other checks to be run
+        # This will be caught later by InstTestClass.test_instrument_test_dates
+        assert ~hasattr(self.test_library.pysat_testing, '_test_dates')
```
Comment (jklenzing) [inline pysat/tests/test_instrument.py:1668]:
```suggestion
        assert not hasattr(self.test_library.pysat_testing, '_test_dates')
```

File: pysat/tests/test_instruments.py [hunk 4]
```diff
@@ -1,16 +1,25 @@
...
+# e.g.,
+# import mypackage
+
 # Import the test classes from pysat
-from pysat.tests.instrument_test_class import generate_instrument_list
+from pysat.utils import generate_instrument_list
 from pysat.tests.instrument_test_class import InstTestClass
 
+
+dir_name = tempfile.mkdtemp()
```
Comment (jklenzing) [inline pysat/tests/test_instruments.py:15]:
```suggestion
```

File: pysat/tests/test_instruments.py [hunk 5]
```diff
@@ -1,15 +1,24 @@
...
```
[the remark this answers is not in the document]
Comment (jklenzing) [reply pysat/tests/test_instruments.py]: left over from one of the changes.  Thanks!

**Comment by jklenzing** [comment]
These should all be working now.  I'm running into a few unrelated bugs in pysatCDAAC and pysatSpaceWeather in the local tests.  One is already documented in pysat/pysatCDAAC#2.  The space weather one is noted in the PR, working on better documentation of this issue now.

**Comment by jklenzing** [comment]
Pinging @aburrell for testing.  Everything should be working in all downstream packages.  The Space Weather bugs seem to be unique to my system (not failing on travis) and are not related to data download.

**Review by aburrell — APPROVED** [review]

**Pull request merged**